[Breaking change] Android: Ti.UI.Color parity

drauggres opened the pull request:
- [x] `Ti.UI.fetchSemanticColor` returns `Ti.UI.Color`
- [x] Support `Ti.UI.Color` instances for components properties
- [x] Automatically change color value on dark/light mode switch

This PR includes commit from #13267 and #13265
Fix #13272 

Read [this](https://github.com/appcelerator/titanium_mobile/pull/13273#issuecomment-1043432465) before merging.

build commented:
<!--
  0 failure: 
  2 warning:  Tests have failed..., There is no linke...
  4 messages
  1 markdown notices
  DangerID: danger-id-Danger;
-->


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>Tests have failed, see below for more information.</td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>There is no linked JIRA ticket in the PR body. Please include the URL of the relevant JIRA ticket. If you need to, you may file a ticket on <a href="https://jira.appcelerator.org/secure/CreateIssue!default.jspa">JIRA</a></td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>:fist: The commits in this PR match our conventions! Feel free to Rebase and Merge this PR when ready.</td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>:x: 1 tests have failed
There are 1 tests failing and 1174 skipped out of 22246 total tests.</td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>:tada: Another contribution from our awesome community member, drauggres! Thanks again for helping us make Titanium SDK better. :thumbsup:</td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-13273/1/artifact/dist/mobilesdk-10.2.0.v20220217055028-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  </tbody>
</table>

### Tests:

<table><tr><th>Classname</th><th>Name</th><th>Time</th><th>Error</th></tr>
<tr><td>android.emulator.main.Titanium.UI.ImageView</td><td>.image with Ti.Blob WebP (12)</td><td>0.076</td><td><pre>Error: expected false to be true
at Assertion.fail (/node_modules/should/cjs/should.js:275:13)
      at Assertion.value (/node_modules/should/cjs/should.js:356:9)
      at ImageView.listener (/ti.ui.imageview.test.js:177:39)
      at ImageView.value (ti:/kroll.js:1604:27)
      at ImageView.value (ti:/kroll.js:1656:25)</pre></td></tr>
</table>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 977ea1f0b4b82f4e5390637354c6e63f9d94a913
</p>

drauggres committed:
feat(android): notify Ti.UI.Color about UI style change

jquick-axway commented:
FYI: The reason we didn't add `Ti.UI.Color` proxy support to Android is because it would be a **HUGE** breaking-change to existing modules since many of them blindly cast a color property's value to `(String)`, which would cause a casting exception.

For example, your changes would break the following...
https://github.com/appcelerator-modules/ti.map/blob/master/android/src/ti/map/CircleProxy.java#L163
https://github.com/appcelerator-modules/ti.map/blob/master/android/src/ti/map/PolygonProxy.java#L122
https://github.com/appcelerator-modules/ti.admob/blob/master/android/src/ti/admob/AdmobView.java#L116
https://github.com/appcelerator-modules/titanium-web-dialog/blob/master/android/src/ti/webdialog/Utils.java#L68

The backward compatible solution that we implemented embeds both dark/light color values into a single string when calling `fetchSemanticColor()`. That's the `"ti.semantic.color:dark=value;light=value"` string format I'm sure you've seen. That's the color string you want to assign to a view proxy's properties so that it will support dark/light theme switching... and this definitely works.

Anyways, that's my 2 cents. This March, you guys can handle this anyway you want. If you want to move forward with this, then my recommendation is to do this in a new Titanium 11.0.0 release and force everyone to rebuild their modules.

drauggres commented:
Thanks @jquick-axway, I'm glad you are still with us.

hansemannn commented:
Very impressive PR! So thrilled to get this reviewed and merged when TiDev can take over.

drauggres committed:
feat(android): parity Ti.UI.Color. allow use in properties

- "Ti.UI.fetchSemanticColor" returns "Ti.UI.Color"
- Support "Ti.UI.Color" instances for components properties

fix #13272

hansemannn commented:
@drauggres A few merge conflicts to resolve here, but no time pressure as it's for 11.0.0 (targeted for September) anyways.

drauggres commented:
> A few merge conflicts to resolve here, but no time pressure as it's for 11.0.0 (targeted for September) anyways.

This happened bacause this PR had to include commits from #13267 and #13265 and they were "merged" with rebase, i.e. not really merged (in git terms) but copies of the commits were created.

I rebased my branch, should be fine now.

hansemannn commented:
@m1ga And this one as well

m1ga committed:
Merge branch 'master' into parity/ti.ui.color_in_props

m1ga committed:
Merge branch 'master' into parity/ti.ui.color_in_props

m1ga committed:
merge conflicts

m1ga committed:
Merge branch 'main' into parity/ti.ui.color_in_props

m1ga committed:
Merge branch 'main' into parity/ti.ui.color_in_props

hansemannn committed:
fix(android): address Ti.UI.Color parity review feedback

- Make UIModule's UIStyleChangedListener registry thread-safe: it is
  touched from the runtime thread (ColorProxy creation) and the UI thread
  (configuration changes). Wrap it in a synchronized map and snapshot the
  listeners before notifying to avoid ConcurrentModificationException.
- Fix TiColorHelper.getDefaultContext() returning null when there is no
  root activity, which could NPE downstream color resolution. Fall back to
  the application context so it is never null.
- Guard ColorProxy re-resolution for name-less colors: a raw color int has
  no resource to re-resolve on a light/dark switch and would otherwise be
  reset to transparent. Only register named colors as style listeners.
- Revert unrelated TiUIButton churn (duplicate imports / reformatting).
- Fix a no-op double ".should" assertion in the getColorResource test.
